File: src/main/java/destiny/core/calendar/decorators/HourDecorator.java
/**
 * @author smallufo 
 * Created on 2008/4/9 at 上午 11:49:59
 */ 
package destiny.core.calendar.decorators;

import com.google.common.collect.ImmutableMap;
import destiny.tools.Decorator;
import destiny.tools.LocaleUtils;
import org.jetbrains.annotations.NotNull;

import java.util.Locale;

public class HourDecorator
{
  private final static ImmutableMap<Locale , Decorator<Integer>> implMap = new ImmutableMap.Builder<Locale , Decorator<Integer>>()
    .put(Locale.TAIWAN  , new HourDecoratorTaiwan())
    .put(Locale.CHINESE , new HourDecoratorChina())
    .put(Locale.ENGLISH , new HourDecoratorEnglish())
    .build();
  
  @NotNull
  public static String getOutputString(int value ,  Locale locale)
  {
    return implMap.get(
      LocaleUtils.getBestMatchingLocale(locale, implMap.keySet()).orElse((Locale) implMap.keySet().toArray()[0])
    ).getOutputString(value);
  }
}

File: src/main/kotlin/destiny/core/calendar/decorators/HourDecorator.kt
/**
 * @author smallufo
 * Created on 2008/4/9 at 上午 11:49:59
 */
package destiny.core.calendar.decorators

import destiny.tools.Decorator
import destiny.tools.getOutputString
import destiny.tools.JSerializable
import java.util.*

object HourDecorator {
  private val implMap = mapOf<Locale, Decorator<Int>>(
    Locale.TAIWAN to HourDecoratorTaiwan(),
    Locale.SIMPLIFIED_CHINESE to HourDecoratorChina(),
    Locale.ENGLISH to HourDecoratorEnglish()
  )

  fun getOutputString(value: Int, locale: Locale): String {
    return implMap.getOutputString(value, locale)
  }
}


class HourDecoratorTaiwan : Decorator<Int>, JSerializable {
  override fun getOutputString(value: Int): String {
    return value.toString() + "時"
  }
}


class HourDecoratorChina : Decorator<Int>, JSerializable {
  override fun getOutputString(value: Int): String {
    return value.toString() + "时"
  }
}

class HourDecoratorEnglish : Decorator<Int>, JSerializable {
  override fun getOutputString(value: Int): String {
    return value.toString()
  }

}
